idaes/tests/test_import.py

```diff
@@ -11,81 +11,285 @@
+                    results[module].timing = float(times[1]) / 1e6
+                    break
+        elif not allow_output:
+            return None
+
+    # Return timing and/or error information
+    return results
+
+
+def collect_import_time(
+    module: str, max_time: float, allow_output: bool = True
+) -> Result:
+    """Import a module by running with 'importtime' option.
+
+    Args:
+        modules: List of absolute name of modules to import (idaes.<something>)
+        max_time: Timeout for process that imports the module
+        allow_output: If False, consider output during import an error; otherwise
+        ignore it.
+
+    Returns:
+        Result object with an error and, if no error, a timing
+    """
+    result, proc = Result(), None
+
+    # Run import in new interpreter with '-X importtime' option
+    import_stmt = f"import idaes.{module}"
+    print(f"import single: '{import_stmt}'")
+    try:
+        proc = subprocess.run(
+            [sys.executable, "-X", "importtime", "-c", import_stmt],
+            stderr=subprocess.STDOUT,
+            stdout=subprocess.PIPE,
+            timeout=max_time,
+        )
+    except subprocess.CalledProcessError as err:  # total failure!
+        result.error = str(err)
+    except subprocess.TimeoutExpired:  # timeout
+        result.error = f"Timeout (> {max_time}s)"
+
+    if proc is not None:
+        output = proc.stdout.decode()
+        if proc.returncode != 0:
+            # find "traceback" text in output
+            result.error = "Non-zero return code:\n"
+            for line in output.splitlines():
+                if not line.startswith("import time:"):
+                    result.add_error(line)
+        else:
+            # Parse output of '-X importtime'
+            prev_import = None
+            for line in output.splitlines():
+                # For rows with timings
+                if line.startswith("import time:"):
+                    # Save if match with current module
+                    if line.endswith(module):
+                        times = re.findall(r"\d+", line)
+                        result.timing = float(times[1]) / 1e6
+                    # In any case, remember module name
+                    prev_import = line[line.rfind("|") + 1:].strip()
+                elif not allow_output:
+                    # Associate dis-allowed output with last seen module name
+                    result.add_error(f"[{prev_import}] {line}")
+
+    # Return timing and/or error information
+    return result
+
+
+####################################################################################
 
 
 @pytest.mark.unit
-def test_import():
-    root_dir = pathlib.Path(idaes.__file__).parent
-    failures, total = importr(root_dir)
-    n = len(failures)
-    assert n == 0, f"{n:d} failures in {total:d} tests: {failures}"
+@pytest.mark.parametrize("module_path", list(idaes_path.rglob("*.py")))
+def test_toplevel_statements(module_path):
```

Comment:
This is more of a general suggestion rather than a comment of this test as it is currently implemented in this PR: should we consider using pylint to perform this check, instead? I can see the following reasons:

- Robustness: I'm personally very wary when it comes to parsing anything, and doubly so if it's source code. The current implementation looks perfectly fine, but I can (conservatively/pessimistically) imagine that there would be edge cases that anything short of a full-blown Python parser wouldn't catch, and that could cause headaches down the line once this is part of the IDAES test suite. Essentially, the functionality here is to extract and count the number of statements in a Python file, which is something that it would IMO be safer and more robust to do at the Abstract Syntax Tree (level). The Python stdlib has an `ast` module, which is the same base of the (`astroid`) module used by pylint
- The way to implement this in pylint would be to create a custom checker. The process is [surprisingly well-documented](http://pylint.pycqa.org/en/latest/how_tos/custom_checkers.html) and, considering we already have all the infrastructure in place for custom Pylint code, it looks like it would be straightforward to do (I'd be happy to take a stab at this)
- Conceptually, having the checks on IDAES modules divided between pytest looking at the *import times* (runtime analysis) and pylint looking at the *code within the module* (static analysis) would better fit the current runtime/static analysis separation of concerns between pytest and pylint which, all else being equal, I think it would make sense to maintain
- As a bonus consideration, building off of a dedicated third-party tool such as pylint (as opposed to adding the module-parsing functionality, which is not specific to IDAES, somewhere within the IDAES test suite) would be more efficient in term of development (and especially maintenance) efforts

Thoughts on this?